c2rust-analyze/src/util.rs

```diff
@@ -115,7 +133,7 @@ pub enum Callee<'tcx> {
     /// Or it could a function in another non-local crate, such as `std`,
     /// as definitions of functions from other crates are not available,
     /// and we definitely can't rewrite them at all.
-    UnknownDef { ty: Ty<'tcx> },
+    UnknownDef(UnknownCall<'tcx>),
```

Comment:
It seems like ideally we'd like to have a separate `Callee` variant for known fns, rather than making them a subset of `UnknownDef`.  The full `AnalysisCtxt` isn't available here, but maybe you could have a `lazy_static` `HashSet` of known fn symbols, which would be initialized from the `KnownFn`s table on first use.

Failing that, maybe it would be cleaner to split out only the specific case we care about (foreign fn, no substs, identified by its symbol name) into a separate `Callee` and leave the rest as the old `UnknownDef(Ty)`.  Having a separate `UnknownCall` currently doesn't buy us much, since the `Indirect`, `Unknown`, and `Direct { is_foreign: false }` cases are still all treated the same everywhere.

Reply:
There are a couple of other cases I was considering:
1. `libc` `fn`s that are in-general, known, but we haven't yet added them to `all_known_fns`.  It'd be nice to have a clear error message in this case that this is a `fn` that should be known.  These would be `Direct { is_foreign: true }`
2. non-local Rust `fn`s, such as those from `std`.  These could be known, though we'd need a different approach to them as they may be generic.  These would be `Direct { is_foreign: false }`.
3. `fn` ptrs that we don't handle yet.  These would be `Indirect` and it'd be clear that they belong to the class of `fn` ptrs, which aren't yet handled (as opposed to the other 2 classes above).
4. Anything else that could get through that we aren't anticipating.  These would be `Unknown`.

Maybe we could embed these within `Callee` rather than having a separate `UnknownCall`, but I think the distinction between the 4 (or 5 if we separate #1 into contained in `all_known_fns` and those not) cases is useful and still pretty simple code.

Reply:
Sounds reasonable to me